chainerrl/agents/ppo.py

```diff
@@ -0,0 +1,265 @@
+            self._flush_last_episode()
+            self.update()
+            self.memory = []
+
+    def _flush_last_episode(self):
+        if self.last_episode:
+            self._compute_teacher()
+            self.memory.extend(self.last_episode)
+            self.last_episode = []
+
+    def _compute_teacher(self):
+        """Estimate state values and advantages of self.last_episode
+
+        TD(lambda) estimation
+        """
+
+        adv = 0.0
+        for transition in reversed(self.last_episode):
+            td_err = (
+                transition['reward']
+                + (self.gamma * transition['nonterminal']
+                   * transition['next_v_pred'])
+                - transition['v_pred']
+                )
+            adv = td_err + self.lambd * adv
+            transition['adv'] = adv
+            transition['v_teacher'] = adv + transition['v_pred']
+
+    def _lossfun(self,
+                 distribs, vs_pred, log_probs,
+                 target_distribs, vs_pred_old, target_log_probs,
+                 advs, vs_teacher):
+        prob_ratio = F.exp(log_probs - target_log_probs)
+        ent = distribs.entropy
+
+        prob_ratio = F.expand_dims(prob_ratio, axis=-1)
+        loss_policy = - F.mean(F.minimum(
+            prob_ratio * advs,
+            F.clip(prob_ratio, 1-self.clip_eps, 1+self.clip_eps) * advs))
+
+        if self.clip_eps_vf is None:
+            loss_value_func = F.mean_squared_error(vs_pred, vs_teacher)
+        else:
+            loss_value_func = F.mean(F.maximum(
+                F.square(vs_pred - vs_teacher),
+                F.square(_F_clip(vs_pred,
+                                 vs_pred_old - self.clip_eps_vf,
+                                 vs_pred_old + self.clip_eps_vf)
+                         - vs_teacher)
+                ))
+
+        loss_entropy = -F.mean(ent)
+
+        # Update stats
+        self.average_loss_policy += (
+            (1 - self.average_loss_decay) *
+            (loss_policy.data - self.average_loss_policy))
+        self.average_loss_value_func += (
+            (1 - self.average_loss_decay) *
+            (loss_value_func.data - self.average_loss_value_func))
+        self.average_loss_entropy += (
+            (1 - self.average_loss_decay) *
+            (loss_entropy.data - self.average_loss_entropy))
+
+        return (
+            loss_policy
+            + self.value_func_coeff * loss_value_func
+            + self.entropy_coeff * loss_entropy
+            )
+
+    def update(self):
+        xp = self.xp
+        target_model = copy.deepcopy(self.model)
+        dataset_iter = chainer.iterators.SerialIterator(
+            self.memory, self.minibatch_size)
+
+        dataset_iter.reset()
+        while dataset_iter.epoch < self.epochs:
+            batch = dataset_iter.__next__()
+            states = batch_states([b['state'] for b in batch], xp, self.phi)
+            actions = xp.array([b['action'] for b in batch], dtype=xp.int32)
```

Comment:
Actions can be np.float32 in continuous-action envs. `, dtype=xp.int32` should be removed.